bin/varnishtest/vtc_tunnel.c

```diff
@@ -0,0 +1,761 @@
+ *
+ * Author: Dridi Boukelmoune <dridi.boukelmoune@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+
+#include <errno.h>
+#include <poll.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "vtc.h"
+
+#include "vtcp.h"
+
+/* SECTION: tunnel tunnel
+ *
+ * The goal of a tunnel is to help control the data transfer between two
+ * parties, for example to trigger socket timeouts in the middle of protocol
+ * frames, without the need to change how both parties are implemented.
+ *
+ * A tunnel accepts a connection and then connects on behalf of the source to
+ * the desired destination. Once both connections are established the tunnel
+ * will transfer bytes unchanged between the source and destination. Transfer
+ * can be interrupted, usually with the help of synchronization methods like
+ * barriers. Once the transfer is paused, it is possible to let a specific
+ * amount of bytes move in either direction.
+ *
+ * SECTION: tunnel.args Arguments
+ *
+ * \-start
+ *        Start the tunnel in background, processing the last given
+ *        specification.
+ *
+ * \-wait
+ *        Block until the thread finishes.
+ *
+ * \-listen STRING
+ *        Dictate the listening socket for the server. STRING is of the form
+ *        "IP PORT", or "HOST PORT".
+ *
+ *        Listens by defaults to a local random port.
+ *
+ * \-connect STRING
+ *        Indicate the server to connect to. STRING is also of the form
+ *        "IP PORT", or "HOST PORT".
+ *
+ *        Connects by default to a varnish instance called ``v1``.
```

Comment:
Left out on purpose, this is a minimal implementation that already served its original purpose! I figured it would be a large-enough change in the first place (and my time budget was limited).

Reply:
sure, no problem!